azure-kusto-data/source/tokenProvider.ts

```diff
@@ -0,0 +1,313 @@
+        }
+        const msiToken = await this.managedIdentityCredential.getToken(this.kustoUri);
+        if (msiToken?.token != null) {
+            return { tokenType: BEARER_TYPE, accessToken: msiToken.token };
+        }
+        throw new Error(`"Failed to obtain MSI token for '${this.kustoUri}' with '${this.clientId}'`)
+    }
+}
+
+/**
+ * AzCli Token Provider obtains a refresh token from the AzCli cache and uses it to authenticate with MSAL
+ */
+export class AzCliTokenProvider extends TokenProviderBase {
+    context = { "authority": this.name };
+    name = "CallbackTokenProvider";
+    azureCliCredentials!: AzureCliCredentials;
+
+    constructor(kustoUri: string) {
+        super(kustoUri);
+    }
+
+    async acquireToken(): Promise<TokenResponse> {
+        if (this.azureCliCredentials == null) {
+            this.azureCliCredentials = await AzureCliCredentials.create({ resource: this.kustoUri });
+        }
+        return this.azureCliCredentials.getToken();
+    }
+}
+
+/**
+ * Acquire a token from MSAL
+ */
+abstract class MsalTokenProvider extends TokenProviderBase {
+    cloudInfo!: CloudInfo;
+    authorityId?: string;
+    initialized: boolean;
+    abstract initClient(): void;
+    abstract acquireMsalToken(): Promise<AuthenticationResult | null>;
+
+    constructor(kustoUri: string, authorityId?: string) {
+        super(kustoUri);
+        this.initialized = false;
+        this.authorityId = authorityId;
+    }
+
+    async acquireToken(): Promise<TokenResponse> {
+        let token;
+        if (!this.initialized) {
+            if (this.cloudInfo == null) {
+                this.cloudInfo = await CloudSettings.getInstance().getCloudInfoForCluster(this.kustoUri);
+                let resourceUri = this.cloudInfo.KustoServiceResourceId;
+                if (this.cloudInfo.LoginMfaRequired) {
+                    resourceUri = resourceUri.replace(".kusto.", ".kustomfa.")
+                }
+                this.scopes = [resourceUri + "/.default"]
+                this.initClient();
+            }
+            this.initialized = true;
+        }
+
+        token = await this.acquireMsalToken();
+        if (token) {
+            return { tokenType: token.tokenType, accessToken: token.accessToken }
+        }
+        throw new Error("Failed to get token from msal");
+    }
+}
+
+/**
+ * Acquire a token from MSAL with username and password
+ */
+export class UserPassTokenProvider extends MsalTokenProvider {
+    name = "UserPassTokenProvider";
+    userName: string;
+    password: string;
+    homeAccountId?: string;
+    msalClient!: PublicClientApplication;
+
+    constructor(kustoUri: string, userName: string, password: string, authorityId?: string) {
+        super(kustoUri, authorityId);
+        this.userName = userName;
+        this.password = password;
+    }
+
+    initClient(): void {
+        const clientConfig = {
+            auth: {
+                clientId: this.cloudInfo.KustoClientAppId,
```

Comment:
This should also come from CloudSettings

Reply:
?
It comes from CloudSettings

Reply:
Apparently going senile.
I think the function below got me confused.